### PR TITLE
feat: short-circuit recall after first Graphiti injection per session

### LINKED ISSUE #168: [NEW-14] Short-circuit ContextEngine recall after first Graphiti injection per session
⚡ **Quick Win**

| Effort | Impact | Category | Origin |
|--------|--------|----------|--------|
| S | 4/5 | integration + performance | novel |

## Summary

Graphiti smart recall currently runs through `ContextEngine.assemble()` on every turn. Under the current plugin-sdk constraints, this means the plugin may call `/get-memory` and emit a new `<graphiti-context>` block for each run.

That creates unnecessary per-turn prompt churn, reduces prompt-cache reuse, and is not aligned with the intended bootstrap-style behavior for Graphiti recall in a session.

Add a deterministic session-scoped short-circuit:

- if this session has **already emitted** a Graphiti context block, skip recall entirely
- if this session has **not emitted** one yet, run `/get-memory` and inject once

## Current behavior

Today `GraphitiContextEngine.assemble()`:

1. converts recent messages into Graphiti message format
2. calls `/get-memory`
3. formats facts into `systemPromptAddition`
4. returns that addition for prompt injection

Because `assemble()` runs per turn, this recall path also runs per turn.

Under the current ContextEngine API, `assemble()` receives:

- `sessionId`
- `sessionKey?`
- `messages`
- `prompt?`
- `model?`

It does **not** receive the already-built effective system prompt.

So the plugin cannot literally do:

```ts
if (systemPrompt.includes("<graphiti-")) return;
```

because that text is not available inside `assemble()`.

## Desired behavior

Graphiti recall should behave like a session bootstrap, not a per-turn reinjection loop.

Desired behavior:

- first eligible turn in a session:
  - call `/get-memory`
  - inject `<graphiti-context>...</graphiti-context>`
  - mark the session as having emitted Graphiti recall
- subsequent turns in the same session:
  - short-circuit immediately
  - return pass-through messages without recall lookup

### 1) Track recall emission state by `sessionId`

Add an in-memory session guard in the plugin lifecycle, e.g.:

```ts
const recalledSessions = new Map<string, true>();
```

or equivalent private state on the engine instance.

### 2) Short-circuit early in `assemble()`

At the start of `assemble()`:

```ts
if (this.recalledSessions.has(params.sessionId)) {
  this.debugLog.log("ce-assemble", { skipped: true, reason: "already_injected" });
  return { messages: params.messages };
}
```

### 3) Mark session only after a real Graphiti block is emitted

If `/get-memory` returns zero facts, do **not** mark the session.

If facts are found and `systemPromptAddition` is returned, mark the session:

```ts
const facts = await this.client.getMemory(graphitiMessages, maxFacts);
if (facts.length === 0) return passThrough;

const systemPromptAddition = formatFactsAsContext(facts);
this.recalledSessions.set(params.sessionId, true);

return {
  messages: params.messages,
  systemPromptAddition,
  estimatedTokens,
};
```

### 4) Keep Graphiti tags in the emitted block

Continue emitting the context inside explicit tags:

```xml
<graphiti-context>
...
</graphiti-context>
```

This preserves future compatibility if OpenClaw later exposes effective system-prompt inspection or placement controls.

### 5) Add a TODO for future SDK support

Document the intended future upgrade path:

```ts
// TODO: if the plugin-sdk exposes effective system-prompt contents or
// placement control for ContextEngine additions, replace this session-scoped
// short-circuit with direct Graphiti-tag detection or append-based placement.
```

## Acceptance criteria

- [ ] `assemble()` skips recall after the first successful Graphiti injection in the same session
- [ ] `/get-memory` is not called on subsequent turns once recall has already been injected for that `sessionId`
- [ ] sessions with zero recalled facts are not incorrectly marked as already injected
- [ ] emitted recall blocks remain wrapped in `<graphiti-context>` tags
- [ ] debug logs clearly show `reason: "already_injected"` when the short-circuit triggers
- [ ] unit test: first turn with facts injects context and marks the session
- [ ] unit test: second turn in same session returns pass-through without calling `/get-memory`
- [ ] unit test: first turn with zero facts does not mark the session
- [ ] unit test: a different `sessionId` still performs recall normally

## Notes

This issue intentionally stays within the current `ContextEngine` contract.

It does **not** depend on moving recall to hooks or on OpenClaw adding append semantics to `systemPromptAddition`.

The short-circuit is implemented via deterministic session state because the current ContextEngine API does not expose the effective system prompt text.

## References

- `ContextEngine.assemble()` params include `sessionId`, `messages`, `prompt`, `model`, but not the effective system prompt
- `systemPromptAddition` is prepend-only in the current ContextEngine flow
- Graphiti smart recall currently uses `/get-memory` and emits `<graphiti-context>` blocks

### PR DESCRIPTION
**Pull request by RobertoGongora**
## Summary

Closes #168

- `assemble()` now skips recall after the first successful Graphiti injection for a given `sessionId`, eliminating per-turn prompt churn and improving prompt-cache reuse
- Lifecycle events (bootstrap, compaction) and deictic references ("continue where we left off") override the gate so recovery still fires when needed
- Sessions with zero recalled facts are not marked, so retry is possible on the next eligible turn

## Changes

- **`context-engine.ts`**: Add `_recalledSessions` Set with session-scoped short-circuit in `assemble()`, explicit clear on `bootstrap()`/`compact()`, and 1000-entry safety cap
- **`test/context-engine.test.ts`**: 6 new tests (inject+mark, no server calls on skip, zero-facts edge case, cross-session isolation, compact override, deictic override) + 1 updated test
- **`CHANGELOG.md`**: Document the new behavior

## Test plan

- [x] All 281 tests pass (`npm test`)
- [x] First turn injects context and marks session
- [x] Second turn short-circuits without calling `/get-memory` or `/search`
- [x] Zero-fact sessions are not incorrectly marked
- [x] Different `sessionId` still performs recall normally
- [x] Compact event overrides the short-circuit
- [x] Deictic reference overrides the short-circuit

### CODE REVIEW

**Review by RobertoGongora — COMMENTED** [review]
## 🔍 Code Review — Kimi K2.5

### Summary
This PR implements session-scoped short-circuit logic in `GraphitiContextEngine.assemble()` to eliminate per-turn prompt churn and improve prompt-cache reuse. After the first successful Graphiti injection per session, subsequent `assemble()` calls skip the expensive `/get-memory` and `/search` operations unless overridden by lifecycle events (bootstrap, compaction) or deictic references. The implementation is clean, well-tested, and includes a 1000-entry safety cap on the tracking Set.

### Findings

#### 🟢 Strengths
- **Solid test coverage**: 6 new tests with clear, descriptive names covering the happy path, zero-facts edge case, cross-session isolation, compact override, and deictic override
- **Conservative safety measures**: 1000-entry cap on `_recalledSessions` prevents unbounded memory growth in long-running processes (context-engine.ts:296)
- **Clean state management**: `bootstrap()` and `compact()` properly clear session state via `_recalledSessions.delete()` (context-engine.ts:388,399,573)
- **Zero-facts handling**: Sessions with empty recall results are not marked, enabling retry on the next eligible turn (implicit via early return at context-engine.ts:252-254 before marking)
- **Minimal invasive changes**: Only 3 files touched with 195 additions and 3 deletions

#### 🟡 Suggestions
- **Test coverage gap — cap clearing**: No test verifies the 1000-entry cap actually clears the Set. While unlikely to hit in practice, a unit test or at least a code path comment would improve confidence. Consider adding a test that mocks 1000 sessions being marked.
- **Potential edge case — session ID collisions**: If `sessionId` values are reused across process restarts, the in-memory Set won't know about prior sessions. This is acceptable for the current use case (per-process optimization), but consider documenting this constraint.

#### 🔴 Critical
None identified.

### Scores

| Category | Score | Notes |
|----------|-------|-------|
| Correctness | 9/10 | Logic is sound, edge cases handled (zero facts, cross-session, overrides) |
| Security | 10/10 | No security concerns; Set lookup is safe, no injection risks |
| Performance | 9/10 | O(1) Set ops, avoids redundant server calls, memory-capped at 1000 |
| Type Safety | 10/10 | TypeScript throughout, no `any` types in changes |
| Test Coverage | 8/10 | 6 new tests cover key paths; missing explicit cap-clearing test |
| Standards | 9/10 | Matches existing patterns, consistent naming, follows TypeScript style |
| Documentation | 10/10 | CHANGELOG.md updated, PR description excellent, inline comments explain the TODO |
| **Overall** | **9.3/10** | |

### Verdict
✅ APPROVE

The implementation is solid, well-tested, and follows project conventions. The short-circuit logic correctly balances performance optimization with safety (lifecycle overrides, zero-facts handling). The 1000-entry cap is a thoughtful touch for production resilience.

**Optional follow-up:** Consider adding a test for the 1000-entry cap clearing behavior, though this is a minor gap given the conservative threshold.

File: context-engine.ts [hunk 1]
```diff
@@ -208,6 +209,17 @@ export class GraphitiContextEngine {
       const gapDetected = isContinuityGap(params.messages.length, { recentEvent: this._lastEvent });
       const deicticDetected = !!lastUserText && hasDeicticReferences(lastUserText);
 
+      // Session-scoped short-circuit: skip recall if this session already received
+      // Graphiti context.  Lifecycle events (bootstrap/compact via _lastEvent) and
+      // deictic references override the gate so recovery still fires when needed.
+      // TODO: if the plugin-sdk exposes effective system-prompt contents or
```
Comment (RobertoGongora) [inline context-engine.ts]: Consider whether the session should be marked as 'recalled' only after a *successful* recall that returns facts, not just any recall attempt. Currently if the Graphiti calls throw, the session is still marked (line 295 runs after the catch). This prevents retry after transient failures.

File: context-engine.ts [hunk 2]
```diff
@@ -281,6 +293,8 @@ export class GraphitiContextEngine {
         ms: Date.now() - start,
       });
 
+      if (this._recalledSessions.size >= 1000) this._recalledSessions.clear();
+      this._recalledSessions.add(params.sessionId);
       return { messages: params.messages, systemPromptAddition, estimatedTokens };
```
Comment (RobertoGongora) [inline context-engine.ts:309]: The 1000-entry cap clears the entire Set. For slightly smoother behavior under heavy load, consider deleting just the oldest entry (using a Map for LRU) rather than wiping all session state at once.

File: context-engine.ts [hunk 3]
```diff
@@ -371,12 +385,14 @@ export class GraphitiContextEngine {
         }]);
 
         this._lastEvent = "compact";
+        this._recalledSessions.delete(params.sessionId);
```
Comment (RobertoGongora) [inline context-engine.ts]: Consider whether `afterTurn` with detected compaction (line 517) should also clear from `_recalledSessions`, since compaction represents a session state change similar to explicit `compact()` calls.

**Review by RobertoGongora — COMMENTED** [review]
## 🔍 Code Review — Kimi (K2.5)

### Summary
This PR implements session-scoped short-circuiting for Graphiti context recall, preventing redundant `/get-memory` and `/search` calls after the first successful injection per session. The implementation is well-structured with thoughtful edge case handling including lifecycle event overrides (bootstrap/compact), deictic reference detection, zero-fact sessions not being marked, and a 1000-entry safety cap on the session tracking Set.

### Findings

#### 🔴 Critical
_None identified._

#### 🟡 Suggestions

- **[context-engine.ts:215-220]** Consider whether `this._lastEvent` should be consumed/cleared immediately when it overrides the short-circuit, rather than waiting until line 228. The current flow works because the flag is read at 216 and consumed at 228, but there's a small window where parallel calls (if they existed) could both see the same event. This is unlikely in practice given the single-threaded nature of the engine lifecycle.

- **[context-engine.ts:298]** The 1000-entry cap clears the entire Set when exceeded. Consider using an LRU pattern or simply deleting the oldest entry before adding a new one when at capacity, to avoid wiping all session state at once. However, given the context (single-user local plugin), this is a minor concern.

- **[context-engine.ts:293-296]** The session marking logic happens *after* the try/catch block that handles the Graphiti calls. If an exception occurs during fact retrieval, the session is still marked as recalled at line 295. Consider whether a failed recall should still mark the session — the current behavior prevents retry storms but also prevents legitimate retries after transient failures.

- **[context-engine.ts:388, 396, 573]** All three lifecycle methods (`compact` ×2, `bootstrap` ×1) clear the session from `_recalledSessions`. This is correct per the PR description, but consider whether this should also happen in `afterTurn` when compaction is detected (line 517), since that's another form of session state change.

#### 🟢 Strengths

- **Solid edge case coverage:** The implementation correctly handles zero-fact sessions (doesn't mark), cross-session isolation (per-sessionId tracking), and explicit overrides (lifecycle events + deictic references).
- **Memory safety:** The 1000-entry cap prevents unbounded growth of the `_recalledSessions` Set in long-running processes.
- **Clean conditional logic:** The short-circuit check at lines 215-220 is readable and well-commented with a TODO for future SDK improvements.
- **Test coverage:** 6 new comprehensive tests covering all major scenarios including the tricky zero-facts edge case and cross-session isolation.
- **Behavioral documentation:** The CHANGELOG entry clearly explains the new behavior, triggers, and overrides.

### Scores

| Category | Score | Notes |
|----------|-------|-------|
| Correctness | 9/10 | Logic is sound, handles edge cases well. Minor consideration around marking session after successful vs attempted recall. |
| Security | 10/10 | No new attack surface; uses existing internal data structures. |
| Performance | 9/10 | Set lookups are O(1). 1000-entry cap prevents memory leaks. Achieves stated goal of reducing redundant API calls. |
| Type Safety | 10/10 | Uses native TypeScript Set<string>, no `any` types introduced. |
| Test Coverage | 9/10 | 6 new tests + 1 updated test cover all major scenarios. Could add test for the 1000-entry cap behavior. |
| Standards | 9/10 | Code style matches existing patterns. Naming conventions consistent. |
| Documentation | 9/10 | CHANGELOG updated, code comments explain the TODO and behavior. |
| **Overall** | **9.3/10** | |

### Verdict
✅ APPROVE

The implementation is solid, well-tested, and addresses the stated issue (#168) effectively. The short-circuit logic with proper override mechanisms will meaningfully reduce token usage and improve prompt cache reuse. Minor suggestions above are optimizations, not blockers.

File: context-engine.ts [hunk 4]
```diff
@@ -208,6 +209,17 @@ export class GraphitiContextEngine {
       const gapDetected = isContinuityGap(params.messages.length, { recentEvent: this._lastEvent });
       const deicticDetected = !!lastUserText && hasDeicticReferences(lastUserText);
```
Comment (RobertoGongora) [inline context-engine.ts:218]: 🟡 **Suggestion**: Consider adding a brief comment explaining why the short-circuit check must happen AFTER `gapDetected` and `deicticDetected` are computed but BEFORE they're used. The ordering matters because deictic detection runs before the early return.

File: context-engine.ts [hunk 5]
```diff
@@ -208,6 +209,17 @@ export class GraphitiContextEngine {
...
 
+      // Session-scoped short-circuit: skip recall if this session already received
+      // Graphiti context.  Lifecycle events (bootstrap/compact via _lastEvent) and
+      // deictic references override the gate so recovery still fires when needed.
+      // TODO: if the plugin-sdk exposes effective system-prompt contents or
+      // placement control for ContextEngine additions, replace this session-scoped
+      // short-circuit with direct Graphiti-tag detection or append-based placement.
+      if (this._recalledSessions.has(params.sessionId) && !this._lastEvent && !deicticDetected) {
```
Comment (RobertoGongora) [inline context-engine.ts:225]: 🟡 **Suggestion**: The TODO comment is helpful. Consider expanding it to mention what the ideal future state would look like (e.g., 'Once plugin-sdk exposes ContextEngine placement control, we could detect existing Graphiti blocks directly instead of using session-scoped tracking').

File: context-engine.ts [hunk 6]
```diff
@@ -281,6 +293,8 @@ export class GraphitiContextEngine {
         ms: Date.now() - start,
       });
 
+      if (this._recalledSessions.size >= 1000) this._recalledSessions.clear();
```
Comment (RobertoGongora) [inline context-engine.ts]: 🟢 **Good**: The 1000-entry cap is a pragmatic safety valve. Consider extracting this to a named constant (e.g., `MAX_RECALLED_SESSIONS = 1000`) for better discoverability and maintainability.

File: test/context-engine.test.ts [hunk 7]
```diff
@@ -1353,6 +1352,177 @@ describe("GraphitiContextEngine", () => {
       expect(lastRequest["/search"]).toBeDefined();
     });
```
Comment (RobertoGongora) [inline test/context-engine.test.ts:1393]: 🟡 **Suggestion**: Instead of `mockOverrides.searchFacts = undefined`, consider explicitly setting to the default mock facts array for clarity: `mockOverrides.searchFacts = SAMPLE_FACTS`. Relying on undefined fallthrough is slightly implicit.

File: test/context-engine.test.ts [hunk 8]
```diff
@@ -1101,13 +1101,12 @@ describe("GraphitiContextEngine", () => {
       expect(result1.systemPromptAddition).toBeDefined();
       expect(result1.systemPromptAddition).toContain("<graphiti-continuity>");
 
-      // Second assemble with few messages but no event — gap by message count still works
+      // Second assemble — session already recalled, short-circuits (bootstrap-once)
```
Comment (RobertoGongora) [inline test/context-engine.test.ts:1143]: 🟢 **Good**: The updated test comment clearly explains the new bootstrap-once behavior. This helps future maintainers understand why the expectation changed from `toBeDefined()` to `toBeUndefined()`.

**Review by RobertoGongora — COMMENTED** [review]
## 🔍 Code Review — GPT

### Summary

This PR implements a session-scoped short-circuit mechanism for Graphiti context recall, eliminating redundant API calls after the first successful injection per session. The implementation is well-considered with proper override conditions for lifecycle events (bootstrap/compact) and deictic references, ensuring recovery still fires when contextually necessary.

### Findings

#### 🔴 Critical
None identified.

#### 🟡 Suggestions

- **[context-engine.ts:221-223]** The 1000-entry cap with automatic clear is a pragmatic safety valve, but consider documenting this behavior more prominently (e.g., in a README or ARCHITECTURE.md) so users understand that long-running processes may occasionally see a "memory reset" behavior after 1000 unique sessions. The inline TODO comment is helpful but could be expanded to explain the trade-off.

- **[context-engine.ts:211-216]** The short-circuit condition checks `!this._lastEvent` but `_lastEvent` is consumed (set to null) immediately after use in lines 227-228. Consider whether there's a race condition possibility if `assemble()` is called concurrently. The current pattern is:
  ```typescript
  const triggerEvent = this._lastEvent;  // line 227
  this._lastEvent = null;                // line 228
  ```
  While likely safe in the single-threaded JavaScript event loop, consider adding a comment about thread-safety assumptions.

- **[test/context-engine.test.ts:1354]** The `zero facts do not mark session` test uses `mockOverrides.searchFacts = undefined` to restore default behavior. Consider explicitly setting to the default mock facts array for clarity rather than relying on `undefined` falling through to default behavior.

#### 🟢 Strengths

- **Clean implementation**: The `_recalledSessions` Set provides O(1) lookup with minimal memory overhead. The 1000-entry cap prevents unbounded growth in long-running processes.

- **Comprehensive test coverage**: Six new tests cover the happy path, server call avoidance, zero-facts edge case, cross-session isolation, compact override, and deictic override. This is excellent coverage for a stateful feature.

- **Proper override logic**: The implementation correctly handles the three override conditions:
  - Lifecycle events (`_lastEvent` being set by bootstrap/compact)
  - Deictic references detected in user text
  - Zero-fact recalls don't mark the session

- **Documentation**: CHANGELOG.md is updated with clear explanation of the new behavior.

- **Backwards compatibility**: The change is additive — existing behavior is preserved when the session hasn't been marked yet.

### Scores

| Category | Score | Notes |
|----------|-------|-------|
| Correctness | 9/10 | Logic is sound, edge cases handled (zero facts, cross-session isolation, overrides) |
| Security | 10/10 | No security concerns; no input validation changes, no secrets exposure |
| Performance | 9/10 | Significant improvement eliminating redundant API calls; memory cap prevents leaks |
| Type Safety | 9/10 | No new `any` usage, types are consistent with existing patterns |
| Test Coverage | 9/10 | 6 comprehensive new tests + 1 updated; covers edge cases well |
| Standards | 8/10 | Follows existing patterns; minor nits on comment clarity |
| Documentation | 8/10 | CHANGELOG updated, inline comments explain rationale; could use more prominent docs |
| **Overall** | **8.9/10** | |

### Verdict

✅ **APPROVE**

This is a well-implemented feature that addresses a real performance concern. The session-scoped short-circuit with proper override conditions demonstrates thoughtful design. The test coverage is comprehensive and the code follows existing patterns. Minor suggestions are non-blocking and can be addressed in follow-up commits if desired.

File: context-engine.ts (hunk 6 above)
Comment (RobertoGongora) [inline context-engine.ts]: 🟡 **Suggestion**: Consider logging when the 1000-entry cap triggers. This would help debug unexpected behavior in long-running server processes.

```typescript
if (this._recalledSessions.size >= 1000) {
  this.debugLog.log("ce-assemble", { action: "recalled_sessions_reset", previousSize: this._recalledSessions.size });
  this._recalledSessions.clear();
}
```

**Review by RobertoGongora — COMMENTED** [review]
## 🔍 Code Review — Gemini (Kimi K2.5)

### Summary

This PR implements session-scoped recall short-circuit for the Graphiti ContextEngine. After the first successful context injection per session, subsequent `assemble()` calls skip recall unless lifecycle events (bootstrap, compaction) or deictic references trigger recovery. This is a smart optimization that reduces prompt churn and improves cache reuse while preserving correctness through well-designed override conditions.

### Findings

#### 🔴 Critical

None. The implementation is logically sound and the edge cases are well-handled.

#### 🟡 Suggestions

- **`context-engine.ts:296`** — The 1000-entry cap clearing mechanism could log when it triggers
  ```typescript
  if (this._recalledSessions.size >= 1000) {
    this.debugLog.log("ce-assemble", { action: "recalled_sessions_reset", previousSize: this._recalledSessions.size });
    this._recalledSessions.clear();
  }
  ```
  This would help debug any unexpected behavior in long-running server processes.

- **`context-engine.ts:219-224`** — The TODO comment mentions replacing the session-scoped short-circuit with direct Graphiti-tag detection. Consider linking this to a GitHub issue for tracking, as this represents future technical debt when the plugin-sdk evolves.

- **`test/context-engine.test.ts:1352-1530`** — The session-scoped short-circuit tests are comprehensive but could benefit from one additional test: verifying that the session *is* marked when facts are returned after a compact event that clears the flag. The current tests verify compact clears the flag and subsequent assembles work, but don't explicitly assert the session gets re-marked after that second recall.

#### 🟢 Strengths

- **Correct override hierarchy**: The short-circuit check (`!this._lastEvent && !deicticDetected`) properly prioritizes explicit user intent (deictic refs) and lifecycle events over the optimization.
- **Zero-facts edge case handled**: Sessions with no recalled facts aren't marked, enabling retry on subsequent turns (line 293-295, tested at line 1419).
- **Cross-session isolation**: The `sessionId`-keyed Set ensures sessions don't interfere with each other.
- **Safety cap**: The 1000-entry limit prevents unbounded memory growth in server deployments.
- **Test quality**: Tests verify both behavior (short-circuit) and mechanism (no server calls), including negative cases.
- **Documentation**: CHANGELOG entry clearly explains the feature, its triggers, and its benefits.

### Scores

| Category | Score | Notes |
|----------|-------|-------|
| Correctness | 9/10 | Logic is sound, edge cases covered. Minor suggestion for re-marking test. |
| Security | 10/10 | No new attack surface; uses existing sessionId. |
| Performance | 9/10 | O(1) Set lookup, eliminates unnecessary HTTP calls. 1000-entry cap is prudent. |
| Type Safety | 10/10 | Proper TypeScript typing throughout. |
| Test Coverage | 9/10 | 6 new tests + 1 updated. Covers main paths and edge cases. Could add re-marking verification. |
| Standards | 9/10 | Follows existing patterns for logging and structure. |
| Documentation | 9/10 | CHANGELOG updated well. TODO could reference an issue. |
| **Overall** | **9.1/10** | |

### Verdict

✅ APPROVE

This is a well-crafted optimization that improves performance without sacrificing correctness. The implementation is clean, thoroughly tested, and properly documented. The short-circuit logic correctly yields to user intent (deictic references) and lifecycle events (bootstrap/compact), ensuring recovery still fires when context loss is detected.

File: context-engine.ts [hunk 9]
```diff
@@ -208,6 +209,17 @@ export class GraphitiContextEngine {
...
       const deicticDetected = !!lastUserText && hasDeicticReferences(lastUserText);
 
+      // Session-scoped short-circuit: skip recall if this session already received
+      // Graphiti context.  Lifecycle events (bootstrap/compact via _lastEvent) and
+      // deictic references override the gate so recovery still fires when needed.
+      // TODO: if the plugin-sdk exposes effective system-prompt contents or
+      // placement control for ContextEngine additions, replace this session-scoped
+      // short-circuit with direct Graphiti-tag detection or append-based placement.
+      if (this._recalledSessions.has(params.sessionId) && !this._lastEvent && !deicticDetected) {
+        this.debugLog.log("ce-assemble", { skipped: true, reason: "already_injected" });
```
Comment (RobertoGongora) [inline context-engine.ts:226]: 🟡 **Suggestion**: The TODO about replacing this with direct Graphiti-tag detection is valuable context. Consider linking to a GitHub issue for tracking, as this represents future technical debt when the plugin-sdk evolves.
